src/config/include/config/Node.hpp

```diff
@@ -0,0 +1,670 @@
+ *
+ * The API for `genny::Node` is strongly inspired by that of `YAML::Node`,
+ * but it provides better error-reporting in the case of invalid
+ * configuration and it allows conversion functions to pass in additional
+ * arguments.
+ *
+ * To convert to non-primitive/built-in types you have two options:
+ *
+ * 1.  Add a constructor to your type that takes a `const Node&` as the
+ *     first parameter. You can pass additional constructor-args in the
+ *     call to `.to<T>()` or `.maybe<T>()`:
+ *
+ *     ```c++
+ *     struct MyFoo {
+ *       MyFoo(const Node& n, int x) {...};
+ *     };
+ *     MyFoo mf = node["foo"].to<MyFoo>(7);
+ *     // calls MyFoo(node["foo"], 7)
+ *     ```
+ *
+ * 2.  Specialize the `genny::NodeConvert` struct for the type:
+ *
+ *     ```c++
+ *     namespace genny {
+ *     template<>
+ *     struct NodeConvert<MyFoo2> {
+ *       // required due to SFINAE
+ *       using type = MyFoo2;
+ *       static type convert(const Node& node, int x) {
+ *         ...
+ *       };
+ *     };
+ *     }  // namespace genny
+ *
+ *     MyFoo2 mf2 = node["foo2"].to<MyFoo2>(7);
+ *     // calls NodeConvert<MyFoo2>::convert(node["foo2"], 7);
+ *     ```
+ *
+ * Whenever possible, prefer the first method of creating a constructor.
+ * This allows you to keep all your logic for how to construct your type
+ * with your type itself.
+ *
+ * Note that it is intentionally impossible to convert `genny::Node` into
+ * `YAML::Node`.
+ */
+class Node {
+public:
+    /**
+     * @param yaml source yaml
+     * @param key string key to associate for this node
+     */
+    Node(const std::string& yaml, std::string key) : Node{parse(yaml), nullptr, std::move(key)} {}
+
+    // explicitly allow copy and move
+    // this is here to protect against regressions
+    // accidentally making this non-copy/move-able
+
+    /**
+     * Copy-construct.
+     */
+    Node(const Node&) = default;
+
+    /**
+     * Move-construct.
+     */
+    Node(Node&&) = default;
+
+    /**
+     * @return a copy of this node.
+     */
+    Node& operator=(const Node&) = default;
+
+    /**
+     * Usage of this node in a moded-from state is undefined.
+     *
+     * @return a moved-to version of this node.
+     */
+    Node& operator=(Node&&) = default;
+
+    /**
+     * What type of node we are.
+     */
+    enum class NodeType {
+        Undefined,
+        Null,
+        Scalar,
+        Sequence,
+        Map,
+    };
+
+    /**
+     * Extract the value via `.to<T>()` if the node
+     * is valid else return the fallback value.
+     *
+     * Deduction allows you to omit the `T` if it matches
+     * the `T` fallback:
+     *
+     * ```c++
+     * auto x = node.value_or(7); // int
+     * auto y = node.value_or(std::string{"foo"}); // std::string
+     *
+     * // or specify it the hard way
+     * auto z = node.value_or<std::string>("foo");
+     * ```
+     *
+     * Like `operator[]` this will "fall-back" to the parent node.
```

Comment:
Whether or not the node "exists" depends on whether the node could be found in the parent:

```c++
    Node node{"seven: 7", ""};
    REQUIRE(node["foo"]["bar"][0]["seven"].value_or(8) == 7);
```

Reply:
I am also kinda confused about the "fallback" behavior. The double fallback behavior to both parent and value_or() is a little bit confusing for generic configNodes (i.e. nodes other than the old actor/phase/workload contexts).

This behavior means that actor authors will have to not reuse any names used by other actors. Otherwise if a workload uses two actors that share the same config name and the author specifies the config in a higher-up context intended for one actor, the other actor is going to inadvertently pick it up.

Fortunately, workloads almost never share configs across actors. But, I do still think genny should internally distinguish between the two types of fallbacks verbally. I think we should use "fallback" to always refer to the value passed to `value_or()` and "inherit" to refer to the parent nodes' values.

E.g. this comment can be rephrased as something like: Like `operation[]`, this will check its chain of parent nodes and inherit the value if it exists anywhere in the chain. If the value is not specified in any parent node, it'll fallback to the value passed to `value_or(8)` (i.e. 8).

Ditto throughout for all usages of fallback/fall-back/"fall back"

Reply:
The other thought I had about this was whether `.value_or()` should behave like `.get_noinherit()` where it won't try to look up the key in the parent if it isn't found in the innermost child.

> Whether or not the node "exists" depends on whether the node could be found in the parent:
> 
> ```c++
>     Node node{"seven: 7", ""};
>     REQUIRE(node["foo"]["bar"][0]["seven"].value_or(8) == 7);
> ```

I find this behavior mystifying - maybe because I've gotten so used to thinking about how path traversal works in MongoDB. What utility do you see in offering this API as opposed to having `node["foo"]["bar"][0]["seven"].value_or(8) == 8`?

Reply:
I don't know if there's a use-case for `get_noinherit` given that you can explicitly ask for `node[".."][key]`. 

> I've gotten so used to thinking about how path traversal works in MongoDB

Can you give an example? I'm only familiar with it in `find()` expressions using `.` to traverse child docs where there is no notion of inheritance.

But I think this issue is only being confused by `value_or` so I removed it. It was intended to just be syntax sugar:

```c++
auto coll = node["Collection"].value_or("coll");
// equivalent to:
auto coll = node["Collection"].maybe<std::string>().value_or("coll");
```

This doesn't address the potential confusion around inheritance though since the same behavior exists without the syntax-sugar:

```c++
    Node node{"seven: 7", ""};
    REQUIRE(node["foo"]["bar"][0]["seven"].maybe<int>().value_or(8) == 7);
```

For in-person review during today's standup:

Given the following YAML:

```yaml
Coll: Test
Phases:
- Doc: foo
- Coll: Bar
- Another:
  - Nested: {Coll: Baz}
```

What should the following return?
What should `node[...].value_or("Or")` return?

```c++
node["Coll"] => Test, Test
node["Phases"]["Coll"] => ?
node["Phases"][0]["Coll"] => 
node["Phases"][1]["Coll"] => Bar, Bar
node["Phases"][2]["Coll"] => ?
node["Phases"][2]["Another"]["Coll"] => ?
node["Phases"][2]["Another"]["Nested"]["Coll"] => Baz, Baz
```

For now I've just killed `value_or` in favor of using vanilla `.maybe<T>().value_or()`.

Reply:
We discussed briefly after standup today. Only thing that I'm deviating from is the decision to have `node["Phases"]["Coll"]` do inheritance to return `Test` rather than barfing due to attempting to access a sequence like a map - this is consistent with what `YAML::Node` does.

Reply:
> I don't know if there's a use-case for `get_noinherit` given that you can explicitly ask for `node[".."][key]`.

But `.get_noinherit()` gave you a definitive answer as to whether the value was specified at that level rather than a higher-level parent node.

> But I think this issue is only being confused by `value_or` so I removed it. It was intended to just be syntax sugar:
> This doesn't address the potential confusion around inheritance though since the same behavior exists without the syntax-sugar:

That's fair. I was thinking `.value_or()` would make people think of `std::optional` and what's being presented is a different kind of API where "missing" means "missing from all levels" and not just the innermost one.

> Can you give an example? I'm only familiar with it in `find()` expressions using `.` to traverse child docs where there is no notion of inheritance.

Hence my confusion. It wasn't specific to the `fallback` parameter per-se and just the inheritance concept I had momentarily blanked out on. Once I reframed it as being tied to the inheritance concept, the rules about the behavior made sense to me.